sysutils/ftp-backup/src/opnsense/mvc/app/library/OPNsense/Backup/FTP.php

```diff
@@ -0,0 +1,244 @@
+    {
+        $cnf = Config::getInstance();
+        $ftp = new FTPSettings();
+        
+        if ($cnf->isValid() && !empty((string)$ftp->enabled)) {
+            $config = $cnf->object();
+
+            $url = (string)$ftp->url;
+            $user = (string)$ftp->user;
+            $password = (string)$ftp->password;
+            $crypto_password = (string)$ftp->password_encryption;
+			$hostname = $config->system->hostname . '.' . $config->system->domain;
+            
+            $configname = 'config-' . $hostname . '-' .  date('Y-m-d_H_i_s') . '.xml';
+            
+            $confdata = file_get_contents('/conf/config.xml');
+            if (!empty($crypto_password)) {
+                $confdata = $this->encrypt($confdata, $crypto_password);
+            }
+            
+			$this->upload_file_content(
+				$url,
+				$user,
+				$password,
+				$configname,
+				$confdata );
+			
+			return $this->listFiles($url, $user, $password);
+        }
+
+    }
+
+    /**
+     * dir listing
+     * @param string $url remote location
+     * @param string $username username
+     * @param string $password password to use
+     * @return array
+     * @throws \Exception when listing files fails
+     */
+    public function listFiles($url, $username, $password)
+    {
+        $curl = curl_init();
+
+        curl_setopt_array($curl, array(
+            CURLOPT_URL => $url,
+            CURLOPT_RETURNTRANSFER => true, // Do not output the data to STDOUT
+            CURLOPT_TIMEOUT => 60,          // maximum time: 1 min
+            CURLOPT_USERPWD => $username . ":" . $password,
+            CURLOPT_DIRLISTONLY => 1,
+        ));
+        
+        $response = curl_exec($curl);
+        
+        $error = curl_error($curl);
+        
+        curl_close($curl);
+        
+        if($error){
+            syslog(LOG_ERR, $error);
+            throw new \Exception($error);
+        }
+
+		$files = explode("\n", $response);
+		
+		// e.g. filter folders ".." and "."
+        return preg_grep('"config-.*\.xml"', $files);
+    }
+
+    /**
+     * upload file
+     * @param string $url remote location
+     * @param string $username remote user
+     * @param string $password password to use
+     * @param string $filename filename to use
+     * @param string $file_content contents to save
+     * @throws \Exception when upload fails
+     */
+    public function upload_file_content($url, $username, $password, $filename, $file_content)
+    {
+        $curl = curl_init();
+    
+        $infile = tmpfile();
+        fwrite($infile, $file_content); 
+        fseek($infile, 0); 
+    
+        curl_setopt_array($curl, array(
+            CURLOPT_URL => $url . $filename,
+            CURLOPT_RETURNTRANSFER => true,
+            CURLOPT_TIMEOUT => 60,
+            CURLOPT_UPLOAD => 1,
+            CURLOPT_INFILE => $infile,
+            CURLOPT_INFILESIZE => strlen($file_content),
+            CURLOPT_USERPWD => $username . ":" . $password,
```

Comment:
There seems to be another interesting option worth to try:
CURLOPT_FTP_CREATE_MISSING_DIRS